jms/src/main/scala/akka/stream/alpakka/jms/JmsProducerStage.scala

```diff
@@ -72,26 +87,21 @@ private[jms] final class JmsProducerStage[A <: JmsMessage, PassThrough](settings
         in,
         new InHandler {
 
-          override def onUpstreamFinish(): Unit = if (inFlightMessagesWithProducer.isEmpty) completeStage()
+          override def onUpstreamFinish(): Unit = if (inFlightMessages.isEmpty) completeStage()
 
           override def onPush(): Unit = {
             val elem: E = grab(in)
             elem match {
               case m: Message[_, _] =>
                 // fetch a jms producer from the pool, and create a holder object to capture the in-flight message.
-                val jmsProducer = jmsProducers.dequeue()
-                val holder = new Holder[E](NotYetThere, futureCB, Some(jmsProducer))
-                inFlightMessagesWithProducer.enqueue(holder)
-
-                // send the element asynchronously, notifying the holder of (successful or failed) completion.
-                Future {
-                  jmsProducer.send(m.message)
-                  elem
-                }.onComplete(holder)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
-              case p: PassThroughMessage[_, _] =>
-                val holder = new Holder[E](NotYetThere, futureCB, None)
-                inFlightMessagesWithProducer.enqueue(holder)
+                val holder = new Holder[E](NotYetThere)
+                inFlightMessages.enqueue(holder)
+                sendWithRetries(SendAttempt[E](m, holder))
+              case _: PassThroughMessage[_, _] =>
+                val holder = new Holder[E](NotYetThere)
+                inFlightMessages.enqueue(holder)
                 holder(Success(elem))
+                pushNextIfPossible()
```

Comment:
As we're already in the graph stage thread,  we can continue with handling the pass-through message directly. Also, the `holder.apply` method got stripped from the `futureCB`, as it got replaced with the `sendCompletedCB` async callback.